### PR TITLE
Check for 3rd-party storage access in ifames

### LINKED ISSUE #14466: Browser-based flows not working properly due to 3rd party cookie policies on Firefox
### Describe the bug

I'd like to set up keycloak server so that the administration console is only available on localhost, behind a reverse proxy.
When I try to access the administration console, after successfully entering admin credentials, the message "Loading the admin console" is displayed and the web page refreshes forever.

For your information, I'm using Firefox 104.0.2 with AdblockPlus (disabling it does not change anything).

I believe this is a bug since my test-case is very simple.
Feel free to ask anything you need to solve the issue :)

### Version

19.0.2

### Expected behavior

We should see the administration console page.
And of course, Keycloak should not log any error.

### Actual behavior

The administration console page loads forever.

On the client-side, a POST request to https://auth.example.com/realms/master/protocol/openid-connect/token fails with 400 Bad Request.
On the server-side, the following is logged:
WARN  [org.keycloak.events] (executor-thread-0) type=REFRESH_TOKEN_ERROR, realmId=c886645d-192c-4e83-94c8-3d414df2311c, clientId=security-admin-console, userId=null, ipAddress=172.18.0.1, error=invalid_token, grant_type=refresh_token, client_auth_method=client-secret

### How to Reproduce?

I created a minimalist test case using docker-compose.
You can then reproduce the issue going to http://localhost:8081/admin/ and logging in (login='admin' password='mdp-admin').
**Note:** I replaced my actual domain name with example.com.

docker-compose.yml file:
```
version: "3"

services:
  keycloak:
    build:
      context: ./conf/keycloak
      dockerfile: Dockerfile-keycloak
    expose:
      - "443"
      - "8080"
    volumes:
      - ./ssl/fullchain.pem:/opt/demo/keycloak/certs/auth.example.com.crt
      - ./ssl/privkey.pem:/opt/demo/keycloak/certs/auth.example.com.key
    environment:
      - KEYCLOAK_ADMIN=admin
      - KEYCLOAK_ADMIN_PASSWORD=mdp-admin
      - KEYCLOAK_FRONTEND_HOSTNAME=auth.example.com
      - KEYCLOAK_FRONTEND_PORT=443
      - KEYCLOAK_ADMIN_URL=http://localhost:8081/
      - KEYCLOAK_HTTPS_CERTIFICATE_FILENAME=auth.example.com.crt
      - KEYCLOAK_HTTPS_CERTIFICATE_KEY_FILENAME=auth.example.com.key
  nginx:
    image: nginx:latest
    ports:
      - "443:443"
      - "8081:8081"
    volumes:
      - ./conf/nginx/nginx.conf:/etc/nginx/conf.d/default.conf
      - ./ssl/fullchain.pem:/etc/nginx/certs/auth.example.com.crt
      - ./ssl/privkey.pem:/etc/nginx/certs/auth.example.com.key
      - ./ssl/chain.pem:/etc/nginx/certs/chain.pem
      - ./ssl/dhparam.pem:/etc/nginx/certs/dhparam.pem
```

Dockerfile-keycloak:
```
FROM quay.io/keycloak/keycloak:19.0.2

ENV KC_HEALTH_ENABLED=true
ENV KC_HTTPS_PORT=443

RUN /opt/keycloak/bin/kc.sh build

ADD . /opt/demo/keycloak/scripts/

WORKDIR /opt/keycloak

ENTRYPOINT [ "/opt/demo/keycloak/scripts/run.sh" ]
```

run.sh:
```
#!/usr/bin/env bash

/opt/keycloak/bin/kc.sh start \
 --hostname="${KEYCLOAK_FRONTEND_HOSTNAME}" \
 --hostname-admin-url=${KEYCLOAK_ADMIN_URL} \
 --proxy=edge \
 --https-certificate-file=/opt/demo/keycloak/certs/${KEYCLOAK_HTTPS_CERTIFICATE_FILENAME} \
 --https-certificate-key-file=/opt/demo/keycloak/certs/${KEYCLOAK_HTTPS_CERTIFICATE_KEY_FILENAME}
```

nginx.conf:
```
server
{
  listen 443 ssl;
  listen [::]:443 ssl;

  server_name auth.example.com;

  ssl_certificate /etc/nginx/certs/auth.example.com.crt;
  ssl_certificate_key /etc/nginx/certs/auth.example.com.key;
  ssl_protocols TLSv1.3;
  ssl_prefer_server_ciphers on;
  ssl_dhparam /etc/nginx/certs/dhparam.pem; 
  ssl_ciphers EECDH+AESGCM:EDH+AESGCM;
  ssl_ecdh_curve secp384r1;
  ssl_session_timeout  10m;
  ssl_session_cache shared:SSL:10m;
  ssl_session_tickets off;
  ssl_stapling on;
  ssl_stapling_verify on;
  ssl_trusted_certificate /etc/nginx/certs/chain.pem;

  # Only expose recommended paths according to documentation:
  # https://www.keycloak.org/server/reverseproxy#_exposed_path_recommendations

  location = / { return 403; }
  location = /health { return 403; }
  location = /metrics { return 403; }
  location = /robots.txt { return 457; }

  error_page 457 = @auth;

  location /
  {
    location ^~ /admin/ { return 403; }
    location ^~ /js/ { return 457; }
    location ^~ /realms/ { return 457; }
    location ^~ /resources/ { return 457; }
    location ^~ /welcome/ { return 403; }
  }

  location @auth
  {
    internal;

    # see /etc/resolv.conf
    resolver 127.0.0.11;
    proxy_pass https://keycloak;

    proxy_ssl_server_name on;

    proxy_set_header Host              $http_host;
    proxy_set_header Upgrade           $http_upgrade;
    proxy_set_header Connection        "upgrade";
    proxy_set_header X-Real-IP         $remote_addr;
    proxy_set_header X-Forwarded-For   $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-Host  $host;
    proxy_set_header X-Forwarded-Port  $server_port;

    proxy_intercept_errors on;
    proxy_http_version 1.1;
    proxy_cache_bypass  $http_upgrade;
    proxy_ssl_protocols TLSv1.3;
  }
}

# http://localhost:8081/: keycloak
server
{
  listen 8081;
  listen [::]:8081;

  location /
  {
    allow 192.168.0.0/16;
    allow 172.16.0.0/12;
    allow 10.0.0.0/8;
    deny all;

    # see /etc/resolv.conf
    resolver 127.0.0.11;
    proxy_pass http://keycloak:8080/;

    proxy_set_header Host              $http_host;
    proxy_set_header Upgrade           $http_upgrade;
    proxy_set_header Connection        "upgrade";
    proxy_set_header X-Real-IP         $remote_addr;
    proxy_set_header X-Forwarded-For   $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-Host  $host;
    proxy_set_header X-Forwarded-Port  $server_port;

    proxy_intercept_errors on;
    proxy_http_version 1.1;
    proxy_cache_bypass  $http_upgrade;
  }
}
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by jonkoops**
Adds additional checks to ensure that browsers that implement [State Partitioning](https://developer.mozilla.org/en-US/docs/Web/Privacy/State_Partitioning) (such as Firefox) can fall back gracefully.

Closes keycloak/keycloak#14466

### CODE REVIEW

File: adapters/oidc/js/static/3p-cookies-step1.html [hunk 1]
```diff
@@ -1,27 +1,26 @@
...
-  ~ See the License for the specific language governing permissions and
-  ~ limitations under the License.
-  -->
-
 <!DOCTYPE html>
 <html>
-<body>
-<script>
-    document.cookie = "KEYCLOAK_3P_COOKIE_SAMESITE=supported; max-age=60; samesite=none; secure"
-    document.cookie = "KEYCLOAK_3P_COOKIE=supported; max-age=60"
```
Comment (jonkoops) [inline adapters/oidc/js/static/3p-cookies-step1.html:23]: Not quite sure why we were setting this other cookie before. @vmuzikar perhaps you can shine a light on this?

Comment (vmuzikar) [reply adapters/oidc/js/static/3p-cookies-step1.html:23]: IIRC, this had to do something with the Apple bug around SameSite that treated `SameSite=None` as `Strict`. Additionally, to support scenarios without TLS (`SameSite=None` requires `Secure`).

Comment (jonkoops) [reply adapters/oidc/js/static/3p-cookies-step1.html:23]: Alright, I am going to add this back to support those edge cases.

**Review by edewit — APPROVED** [review]

**Comment by mposolda** [comment]
One note: in the other PR https://github.com/keycloak/keycloak/pull/14560 we discussed to change the format of KEYCLOAK_SESSION cookie to contain only "session-id" instead of the triplet "realm/user-id/user-session-id" . Not sure whether to take this PR as an opportunity to change this right away or whether to rather use follow-up PR?

Another thing is, that all 3 files in this directory https://github.com/keycloak/keycloak/tree/main/adapters/oidc/js/static likely should be in different place as they are not part of the keycloak-js adapter itself, so it is the question why there are inside codebase in the keycloak-js adapter codebase. Changing this is likely out-of-scope of this PR, but just fyi :-)

**Comment by jonkoops** [comment]
> Not sure whether to take this PR as an opportunity to change this right away or whether to rather use follow-up PR?

Since this fixes a rather critical issue I'd prefer to keep the refactoring to a later point. I have some work ready to improve the code quality here significantly as well, however this is blocked on the removal of PhantomJS (see #14702).

> The question why there are inside codebase in the keycloak-js adapter codebase.

This has been so historically, but this will change. @stianst has done the work for this in #14348.

**Comment by mposolda** [comment]
@jonkoops Thanks for the info. Makes sense to do that as a follow-up. Created https://github.com/keycloak/keycloak/issues/14731 for the KEYCLOAK_SESSION cookie change

**Review by vmuzikar — COMMENTED** [review]
@jonkoops Thanks for the fix, nice work! Changes LGTM (although with my poor JS skillset it took me a while to navigate through all the changes from which most of them are IMHO a bit out of scope of the issue, like e.g. formatting changes or single to double quotes swaps 🙂).

I was also able to verify the changes fixed this super easy reproducer:
```
./kc.sh start-dev --hostname=example.com --hostname-admin=localhost
```

What worries me a bit is how do we automatically test this? Especially the corner cases (storage API is not present, etc.). Do we already have an issue created for this? Since it is a critical issue, we can address the testing as a follow-up.

One more thing is that we should run the existing cookies tests in the internal pipeline. If you have already run them, please ignore this note. 🙂

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@jonkoops One last thing. Can you please link this to a GH Issue?

**Review by vmuzikar — APPROVED** [review]

**Comment by jonkoops** [comment]
> What worries me a bit is how do we automatically test this? Especially the corner cases (storage API is not present, etc.). Do we already have an issue created for this? Since it is a critical issue, we can address the testing as a follow-up.

This is indeed a problem and quite hard to test this scenario as we'd have to run a reverse proxy, or somehow trick the browser into triggering the behavior. Created #14761 for this.

**Comment by cscocca** [comment]
Why the 26.2.5 version is not present on [npmjs](https://www.npmjs.com/package/keycloak-js?activeTab=versions) site?

**Pull request merged**